src/mast/mod.rs

```diff
@@ -417,26 +421,45 @@ fn monomorphize_expr<B: Backend>(
                 .expect("function not found")
                 .to_owned();
 
-            // monomorphize the function call
-            let (fn_info_mono, typ) = instantiate_fn_call(ctx, fn_info, &observed, expr.span)?;
-
             let args_mono = observed.clone().into_iter().map(|e| e.expr).collect();
 
-            let fn_name_mono = &fn_info_mono.sig().name;
-            let mexpr = Expr {
-                kind: ExprKind::FnCall {
-                    module: module.clone(),
-                    fn_name: fn_name_mono.clone(),
-                    args: args_mono,
-                },
-                ..expr.clone()
-            };
+            // check if this function is already monomorphized
+            if ctx.functions_instantiated.contains_key(&old_qualified) {
```

Comment:
Shouldn't this be checked to contain the new signature as a key? I think in the code below the `zip` function is monomorphized two times, even if the LEN parameter is the same. Though, maybe I'm missing the point of the PR :smile: 
```rust
fn zip(a1: [Field; LEN], a2: [Field; LEN]) -> [[Field; 2]; LEN] {
    let mut result = [[0; 2]; LEN];
    for index in 0..LEN {
        result[index] = [a1[index], a2[index]];
    }
    return result;
}

fn main(pub arr: [Field; 3]) -> Field {
    let expected = [1, 2, 3];
    for pair in zip(arr, expected) {
        assert_eq(pair[0], pair[1]);
    }
    for pair in zip(arr, expected) {
        assert_eq(pair[0], pair[1]);
    }

    return arr[0];
}
```

Reply:
Oh yeah, indeed. It should always check the monomorphized name.

I have just pushed a fix that also includes a refactor to better encapsulate the resolutions for the generic values and types. This refactor makes it always resolves the types so as to get the monomorphized qualified name, without having to re-instantiate a function.